mission_control/navigator_missions/nav_missions/circle_tower.py

```diff
@@ -7,55 +7,114 @@
 from sensor_msgs.msg import PointCloud
 import datetime
 
-@txros.util.cancellableInlineCallbacks
-def main(navigator):
-    res = navigator.fetch_result()
-
-    buoy_field = yield navigator.database_query("BuoyField")
-    buoy_field_point = navigator_tools.point_to_numpy(buoy_field.objects[0])
 
-    #yield navigator.move.set_position(buoy_field_point).go()
+BF_WIDTH = 60.0  # m
+BF_EST_COFIDENCE = 10.0  # How percisly can they place the waypoints? (m)
+TOTEM_SAFE_DIST = 7  # How close do we go to the totem
+ROT_SAFE_DIST = 3.5  # How close to rotate around it
 
-    circle_colors = ['blue', 'red']
-    color_map = {'blue': (255, 0, 0), 'red': (0, 0, 255)}
+@txros.util.cancellableInlineCallbacks
+def main(navigator, **kwargs):
+    # rgb color map to param vlues
+    color_map = {'BLUE': [0, 0, 1], 'RED': [1, 0, 0], 'YELLOW': [1, 1, 0], 'GREEN': [0, 1, 0]}
 
     explored_ids = []
     all_found = False
 
-    while not all_found:
-        target_totem, explored_ids = yield get_closest_buoy(navigator, explored_ids)
+    # Get colors of intrest and directions
+    c1 = navigator.mission_params['totem_color_1'].get()
+    d1 = navigator.mission_params['totem_direction_1'].get()
+    c2 = navigator.mission_params['totem_color_2'].get()
+    d2 = navigator.mission_params['totem_direction_2'].get()
+    
+    colors = [c1, c2]
+    directions = [d1, d2]
 
-        if target_totem is None:
+    buoy_field = yield navigator.database_query("BuoyField")
+    buoy_field_point = navigator_tools.point_to_numpy(buoy_field.objects[0].position)
+
+    _dist_from_bf = lambda pt: np.linalg.norm(buoy_field_point - pt)
+
+    # We want to go to an observation point based on solar position
+    # TODO: Check which side to go to based on relative distance
+    center = navigator.move.set_position(buoy_field_point).set_orientation(get_solar_q())
+    obs_point = center.backward(BF_WIDTH / 2 + BF_EST_COFIDENCE) 
+    yield obs_point.go()
+
+    # Next jig around to see buoys, first enforce that we're looking at the buoy field
+    buoy_field_point[2] = 1
+    #yield navigator.move.look_at(buoy_field_point).go(move_type='skid', focus=buoy_field_point)
+    yield navigator.nh.sleep(3)
+    #yield navigator.move.yaw_right(.5).go(move_type='skid')
+    #yield navigator.nh.sleep(3)
+    #yield navigator.move.yaw_left(1).go(move_type='skid')
+
+    
+    # TODO: What if we don't see the colors?
+    for color, direction in zip(colors, directions):
+        color = yield color
+        direction = yield direction
+
+        target = yield get_colored_buoy(navigator, color_map[color])
+        if target is None or _dist_from_bf(navigator_tools.point_to_numpy(target.position)) > (BF_WIDTH / 2 + BF_EST_COFIDENCE):
+            # Need to do something
             fprint("No suitable totems found.", msg_color='red', title="CIRCLE_TOTEM")
```

Comment:
Rules: ```Entering  the obstacle avoidance field and avoiding at least one   obstacle   or   circling   any   totem   will   be considered as an attempt at completing this task```

So we should definitely have a case to just circle a random totem and get out if we can't get the colors we want. Maybe even if there are no totems just drive to the buoy field and drive out and let rrt get us some points.